### ISSUE TITLE
Update app configuration but NOT trigger a restart (deployment)

### ISSUE THREAD

**Issue by feisuzhu** [mesosphere/marathon#1635]
We have the following use case:
1. We are running Clojure apps that can apply hot code update, and when performing a cold restart, the app should be restarted at a specific time to minimize impact. After hot code update, we will publish a new version to include applied patch. Due to this the app configuration is out of sync with the current state, effectively lost the ability to scale with a single click.
2. We are trying to managed caches (Redis/Memcached) with Marathon, in this case, the stale configuration is acceptable but restarting is not.  e.g.:  We would change cache memory from 1GB to 2GB, but keep existing 1GB instances work with newly spawned 2GB instances, restarting them are costly.
3. We put some metadata in apps' labels, e.g.: nginx upstream conf. When changing these metadata, (adding/remove an 'ip_hash' directive for example), restarting is not desired at all.

**Comment by bobrik** [comment]
1. Run another app for a new version, add it to LB/SD, remove previous app from LB/SD. Kill old app when you're ready. Alternatively, just don't do anything related with marathon, just tell your apps to reload the code or whatever mechanism clojure provides.
2. Run bigger caches in another app, add it to LB or SD.
3. Labels are propagated to mesos slaves, marathon must spawn new instances to reflect the change there.

**Comment by feisuzhu** [comment]
>Run another app for a new version, add it to LB/SD, remove previous app from LB/SD. Kill old app when you're ready.

It's too much human interaction, and error prone.
Changing configuration of LB/SD for every deployment also defeats the purpose of LB/SD.

>Alternatively, just don't do anything related with marathon, just tell your apps to reload the code or whatever mechanism clojure provides.

This is what we are doing now. But you can't spawn new instances since configuration is stale. We have to re-apply patches to newly spawned instances. Again, too much human interaction.

> Run bigger caches in another app, add it to LB or SD.

Same app, same role, but multiple apps?
This will confuse others and be bitten in obscure ways.
```
/cache
/cache-2gb-20150601
/cache-2gb-no-bgsave-20150602
/cache-2gb-actually-no-bgsave-is-a-bad-idea-20150603
......
```

> Labels are propagated to mesos slaves, marathon must spawn new instances to reflect the change there.

Fair enough.
We always query labels from marathon, so this is not a problem for us.

**Comment by bobrik** [comment]
> It's too much human interaction, and error prone.

It can be automated, no need to use humans. I run debug versions of my app alongside production version and it is picked up automatically by LB (check out [zoidberg](https://github.com/bobrik/zoidberg)).

> Changing configuration of SD for every deployment also defeats the purpose of SD.

That's the purpose of SD, isn't it? You change it on every deployment since new instances are added and old ones are removed. I might miss your point here.

> This is what we are doing now. But you can't spawn new instances since configuration is stale. We have to re-apply patches to newly spawned instances. Again, too much human interaction.

You need 2 files to deploy:

1. Fully assembled asset that is used when new instances are spawned.
2. Patch from original asset to the final one.

To update code without respawning instances you apply the patch. To scale the number of instances you use full asset that is already patched.

> Same app, same role, but multiple apps?
> This will confuse others and be bitten in obscure ways.

Having multiple running configurations for the same app is also confusing. I don't think you should have a lot of versions running simultaneously, they have to converge to one at some point.

**Comment by feisuzhu** [comment]
>> Changing configuration of SD for every deployment also defeats the purpose of SD.

>That's the purpose of SD, isn't it? You change it on every deployment since new instances are added and old ones are removed. I might miss your point here.

Sorry…… I mean, changing configuration __manually__.
A new app in marathon is also a new app to SD. I have to add the new app name to LB. Wait... where is the 'discover' part of SD?
Or to automate this, I have to trick SD to believe "these two(or a dozen) apps are actually the same app, put them in a same entry", which sounds wrong.

>> This is what we are doing now. But you can't spawn new instances since configuration is stale. We 

>have to re-apply patches to newly spawned instances. Again, too much human interaction.
You need 2 files to deploy:
Fully assembled asset that is used when new instances are spawned.
Patch from original asset to the final one.
To update code without respawning instances you apply the patch. To scale the number of instances you use full asset that is already patched.

It's not about respawning, it about scale the number of instances, and no, I can't do this since I can't change the configuration to point to newly assembled assets (or marathon will trigger a full restart which are not acceptable). I can't just add a new app since there are constraints like hostname:UNIQUE can't be violated.

>> Same app, same role, but multiple apps?
This will confuse others and be bitten in obscure ways.

> Having multiple running configurations for the same app is also confusing. I don't think you should have a lot of versions running simultaneously, they have to converge to one at some point.

Every task has it's version info attached, some sign of "this task is not running on current version" on UI will do the work. 
Actually this is even needed without the feature I requested, half way aborted deployments will cause this too.

**Comment by feisuzhu** [comment]
I'll look into the zoidberg and see if it helps.

**Comment by aquamatthias** [comment]
Every change to an application definition implies a restart, as long as there are running instances in the cluster. There is currently no clean way of applying the changed settings (memory, cpu shares, env variables etc) other than restarting the process.

**Comment by feisuzhu** [comment]
@aquamatthias 
I'm not asking for the ability to change settings without restarting,
I'm asking for the ability of 'just not touching those old processes'.

**Comment by aquamatthias** [comment]
Hey @feisuzhu 

The title of this ticket suggests, that you want to update an application definition, but not restarting existing processes. Since configuration changes can deal with cgroups changes (changed resource settings like cpu, mem, disk etc) executor changes (roles, ports etc.) and application changes (version, config, environment) we restart on the level of Mesos Tasks to apply all settings. 

Marathon enforces configuration changes to all instances of one application definition, since that is the only way to prevent a stale configuration. One solution could be to add a flag to override this behavior. This interferes with other features like service discovery, rolling updates etc. I will reopen this ticket to see, if there is a demand for this feature.

Things to consider:
- you can define a minimum health capacity to define how the upgrade process should be performed
- you can use different application definitions (e.g. under the same group) to achieve this goal
- take a look at chronos to be able to schedule a time for the update

**Comment by dpgeekzero** [comment]
My use case for this:

I have dedicated servers, each running an instance of a docker container.  When there is a code update I would like to enable my CI service to issue a PUT to marathon, updating the config to point to the newest production build.

However, I can't safely restart my service without disconnecting clients.  If the config was updated then I could use some external script to wait for a moment when all clients are disconnected and restart the service.

Over a day or so, all of the instances of my task would update.
